tika-parsers/src/main/java/org/apache/tika/parser/html/charsetdetector/StandardHtmlEncodingDetector.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.tika.parser.html.charsetdetector;
+
+import org.apache.commons.io.input.BoundedInputStream;
+import org.apache.tika.detect.CompositeEncodingDetector;
+import org.apache.tika.detect.EncodingDetector;
+import org.apache.tika.metadata.Metadata;
+import org.apache.tika.mime.MediaType;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+
+import static java.util.Arrays.asList;
+import static org.apache.tika.parser.html.charsetdetector.CharsetAliases.getCharsetByLabel;
+
+/**
+ * An encoding detector that tries to respect the spirit of the HTML spec
+ * part 12.2.3 "The input byte stream", or at least the part that is compatible with
+ * the implementation of tika.
+ * <p>
+ * https://html.spec.whatwg.org/multipage/parsing.html#the-input-byte-stream
+ * <p>
+ * If a resource was fetched over HTTP, then HTTP headers should be added to tika metadata
+ * when using {@link #detect}, especially {@link Metadata.CONTENT_TYPE}, as it may contain charset information.
+ * <p>
+ * This encoding detector may return null if no encoding is detected.
+ * It is meant to be used inside a {@link org.apache.tika.detect.CompositeDetector}.
+ * For instance:
+ * <pre> {@code
+ *     EncodingDetector detector = new CompositeDetector(
+ *         new StandardHtmlEncodingDetector(),
+ *         new Icu4jEncodingDetector()
+ *     );
+ * }</pre>
+ * <p>
+ * This detector is stateless, and the same instance can be used on several different input streams.
+ */
+public final class StandardHtmlEncodingDetector implements EncodingDetector {
+
+    /**
+     * A composite encoding detector chaining a {@link StandardHtmlEncodingDetector}
+     * (that may return null) and a {@link StandardIcu4JEncodingDetector} (that always return a value)
+     * This full thus always returns an encoding, and still works very well with data coming
+     * from the web.
+     */
+    public static final EncodingDetector FULL_DETECTOR = new CompositeEncodingDetector(asList(
+            new StandardHtmlEncodingDetector(),
+            new StandardIcu4JEncodingDetector()
+    ));
+    private static final int META_TAG_BUFFER_SIZE = 8192;
```

Comment:
Please add the instance-level `markLimit` like we do in HtmlEncodingDetector as a \@Field so that users can configure a different markLimit.

Reply:
I did this in https://github.com/apache/tika/pull/242/commits/e7cda261f6aa2eb0102a8ffb53e454a4f3655bdd

However, this makes the detector mutable, and makes things more complex. Couldn't we set  the `markLimit` in a constructor parameter, in order to keep the detector immutable ?